pkg/get/get.go

```diff
@@ -0,0 +1,194 @@
+	Repo           string
+	Owner          string
+	Version        string
+	URLTemplate    string
+	BinaryTemplate string
+}
+
+func (t Tool) Latest() bool {
+	return len(t.Version) == 0
+}
+
+func (tool Tool) GetURL(os, arch, version string) (string, error) {
+	if len(tool.URLTemplate) == 0 {
+		if len(version) == 0 {
+			releases := fmt.Sprintf("https://github.com/%s/%s/releases/latest", tool.Owner, tool.Name)
+			var err error
+			version, err = findGitHubRelease(releases)
+			if err != nil {
+				return "", err
+			}
+		}
+
+		var err error
+		t := template.New(tool.Name + "binary")
+		var funcs = map[string]interface{}{"HasPrefix": func(s, prefix string) bool { return strings.HasPrefix(s, prefix) }}
+
+		t = t.Funcs(funcs)
+		t, err = t.Parse(tool.BinaryTemplate)
+		if err != nil {
+			return "", err
+		}
+
+		var buf bytes.Buffer
+		pref := map[string]string{
+			"OS":   os,
+			"Arch": arch,
+			"Name": tool.Name,
+		}
+		fmt.Println(pref)
+		err = t.Execute(&buf, pref)
+		if err != nil {
+			return "", err
+		}
+		res := buf.String()
+		fmt.Println(res)
+		return fmt.Sprintf(
+			"https://github.com/%s/%s/releases/download/%s/%s",
+			tool.Owner, tool.Name, version, res), nil
+	}
+
+	var err error
+	t := template.New(tool.Name)
+	t, err = t.Parse(tool.URLTemplate)
+	if err != nil {
+		return "", err
+	}
+
+	var buf bytes.Buffer
+	err = t.Execute(&buf, map[string]string{
+		"OS":      os,
+		"Arch":    arch,
+		"Version": version,
+	})
+	if err != nil {
+		return "", err
+	}
+	res := buf.String()
+	return res, nil
+}
+
+func findGitHubRelease(url string) (string, error) {
+	timeout := time.Second * 5
+	client := makeHTTPClient(&timeout, false)
+	client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+		return http.ErrUseLastResponse
+	}
+
+	req, err := http.NewRequest(http.MethodHead, url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	res, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+	if res.StatusCode != 302 {
+		return "", fmt.Errorf("incorrect status code: %d", res.StatusCode)
+	}
+
+	loc := res.Header.Get("Location")
+	if len(loc) == 0 {
+		return "", fmt.Errorf("unable to determine release of kubeseal")
+	}
+	version := loc[strings.LastIndex(loc, "/")+1:]
+	return version, nil
+}
+
+func MakeTools() []Tool {
+	tools := []Tool{
+		{
+			Owner: "openfaas",
+			Repo:  "faas-cli",
+			Name:  "faas-cli",
+			BinaryTemplate: `{{ if HasPrefix .OS "ming" -}}
+{{.Name}}.exe
+{{- else if eq .OS "darwin" -}}
+{{.Name}}-darwin
+{{- else if eq .Arch "armv6l" -}}
+{{.Name}}-armhf
+{{- else if eq .Arch "armv7l" -}}
+{{.Name}}-armhf
+{{- else if eq .Arch "aarch64" -}}
+{{.Name}}-arm64
+{{- end -}}`,
+		},
+		{
+			Owner:       "kubernetes",
+			Repo:        "kubernetes",
+			Name:        "kubectl",
+			Version:     "v1.18.0",
+			URLTemplate: `https://storage.googleapis.com/kubernetes-release/release/{{.Version}}/bin/{{.OS}}/{{.Arch}}/kubectl`,
```

Comment:
This doesn't work so well with building arbitrary URLs - I need to run a function to determine the Arch value i.e. `amd64` for 64-bit MacOS. I am now thinking about having a `func` instead.

Originally the idea of a template was to be able to have this structure on disk and not compiled-in, using a `func` would prevent that.

Reply:
Seems like you can have a more complex template i.e.

```
`{{$arch := "arm"}}
{{- if eq .Arch "x86_64" -}}
{{$arch = "amd64"}}
{{- end -}}
			https://storage.googleapis.com/kubernetes-release/release/{{.Version}}/bin/{{.OS}}/{{$arch}}/kubectl`
```

cc @LucasRoesler @Waterdrips

Reply:
if you want it on disk, then i think using an gotmpl input is required. The simplest thing is to probably just pass the Tool object in as the template context